cirq-core/cirq/experiments/z_phase_calibration.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provides a method to do z-phase calibration for excitation-preserving gates."""
+from typing import Union, Optional, Sequence, Tuple, Dict, TYPE_CHECKING
+
+import multiprocessing
+import multiprocessing.pool
+import matplotlib.pyplot as plt
+import numpy as np
+
+from cirq.experiments import xeb_fitting
+from cirq.experiments.two_qubit_xeb import parallel_xeb_workflow
+from cirq import ops
+
+if TYPE_CHECKING:
+    import cirq
+    import pandas as pd
+
+
+def z_phase_calibration_workflow(
+    sampler: 'cirq.Sampler',
+    qubits: Optional[Sequence['cirq.GridQubit']] = None,
+    two_qubit_gate: 'cirq.Gate' = ops.CZ,
+    options: Optional[xeb_fitting.XEBPhasedFSimCharacterizationOptions] = None,
+    n_repetitions: int = 10**4,
+    n_combinations: int = 10,
+    n_circuits: int = 20,
+    cycle_depths: Sequence[int] = tuple(np.arange(3, 100, 20)),
+    random_state: 'cirq.RANDOM_STATE_OR_SEED_LIKE' = None,
+    atol: float = 1e-3,
+    pool_or_num_workers: Optional[Union[int, 'multiprocessing.pool.Pool']] = None,
```

Comment:
Please make a single-process execution the default behavior of the pool argument.
Running via multiprocessing has previously caused unexpected problems in #6674.
It is better if users explicitly ask for multiprocessing so there is a better chance that
potential bugs appear then rather as a default behavior.

I would also recommend to split the `pool_or_num_workers` argument into two exclusive arguments, `pool` and `num_workers: Optional[int]`.
The invocation via `num_workers` could then use a recursive call as

```
with multiprocessing.Pool(num_workers) as pool:
    return z_phase_calibration_workflow(sampler, ..., pool=pool)
```

which would be more robust for closing the pool in case of exception, you'd also won't need to track a `local_pool` flag.

Reply:
> Please make a single-process execution the default behavior of the pool argument
this was the initial behaviour but @eliottrosenberg requested the default be multiprocessing

> I would also recommend to split the pool_or_num_workers argument into two exclusive arguments, pool and num_workers: Optional[int]

I condidered this but I thought the api would be a bit convoluted

Reply:
@pavoljuhas I think that most users would use it with the default values without passing a parallel pool, and it is very slow that way.

Reply:
My concern is all other functions with `pool`-like argument interpret `pool=None` as *run in series*.
If reading `z_phase_calibration_workflow(..., pool_or_num_workers=None)`  my expectation would be to have no pool and no workers, but in fact this starts a pool of ncpu size, which may cause memory problems on hosts with many CPUs.

Can we instead use some reasonable default, say 4, and not accept `None`, ie, make that argument of `Union[int, 'multiprocessing.pool.Pool']` type?

Also renaming to `num_workers_or_pool` might give a better hint to pass integer and make distinction from pool-like arguments.

Reply:
done